RFC.md

```diff
@@ -0,0 +1,466 @@
+
+```python
+class DeepFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		...
+	):
+
+	def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value caches for attention calculation.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.decoder.setup_caches(batch_size, dtype)
+
+	def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.decoder.caches_are_enabled()
+
+	def reset_caches(self):
+        """Reset the key value caches."""
+        self.decoder.reset_caches()
+
+	def forward(
+		self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+	if 	encoder_input is not None:
+		# during decoding, the output of the encoder
+		# is cached in the decoder cross attention kv_cache.
+		# so encoder_input is only called when there's a new image
+		encoder_embed = self.encoder(**encoder_input)
+	# input_pos is tracked by kv_cache now, but left in for
+	# now to support sample packing and can be used here
+	if input_pos is not None and encoder_mask is not None:
+		encoder_mask = encoder_mask[input_pos]
+	output = self.decoder(tokens, mask, encoder_embed, encoder_mask, input_pos)
+	return output
+
+ ```
+
+EarlyFusion is more complicated from the fusion side, but does not require a special Decoder. EarlyFusion needs to call `decoder.tok_embeddings` directly so it can merge the embeddings with the encoded embeddings. As to not modify the `state_dict` it employs two hooks to keep `tok_embeddings` inside the decoder for the `state_dict`.
+
+```python
+class EarlyFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		encoder_token: int,
+	):
+		self.decoder = decoder
+		self.encoder = encoder
+		self.encoder_token = encoder_token
+
+		...
+
+		# A little surgery in the decoder to give the
+		# fusion module access to control the embeddings
+		# The alternative is to pass a special tok_embeddings
+		# module into TransformerDecoder builder that does the
+		# merging there
+		self.tok_embeddings = decoder.tok_embeddings
+		decoder.tok_embeddings = nn.Identity()
+
+		self.register_state_dict_post_hook(Test._state_dict_hook)
+        # TODO: Switch to register_load_state_dict_pre_hook in v2.5
+        self._register_load_state_dict_pre_hook(
+	        Test._load_state_dict_hook,
+	        with_module=True
+	    )
```

Comment:
no idea what this is doing :)

Reply:
Since FusionLayer has to assign the original layer to a variable ("layer"), this modifies the state dict key by prepending an extra "layer." term to those parameters. These two state_dict hooks remove and add the "layer." when outputting and loading the state_dict respectively.